File: backend/apps/modes/modes.py
import json
import os
import logging
from contextlib import asynccontextmanager
from fastapi import HTTPException
from backend.config.Apps import SubApp
from backend.apps.common.json_store import JsonStore
from backend.apps.modes.models import Mode, ModeCreate, ModeUpdate
from backend.apps.modes.builtin import BUILTIN_MODES

logger = logging.getLogger(__name__)

from backend.config.paths import MODES_DIR as DATA_DIR


@asynccontextmanager
async def modes_lifespan():
    os.makedirs(DATA_DIR, exist_ok=True)
    for builtin in BUILTIN_MODES:
        path = os.path.join(DATA_DIR, f"{builtin.id}.json")
        if not os.path.exists(path):
            _save(builtin)
    yield


modes = SubApp("modes", modes_lifespan)


_store = JsonStore(Mode, DATA_DIR, not_found_detail="Mode not found")

_load_all = _store.load_all
_save = _store.save
_load = _store.load
load_mode = _store.load_or_none


@modes.router.get("/list")
async def list_modes():
    builtin_defaults = {m.id: m.model_dump() for m in BUILTIN_MODES}
    return {"modes": [m.model_dump() for m in _load_all()], "builtin_defaults": builtin_defaults}


@modes.router.get("/{mode_id}")
async def get_mode(mode_id: str):
    return _load(mode_id).model_dump()


@modes.router.post("/create")
async def create_mode(body: ModeCreate):
    mode = Mode(
        name=body.name,
        description=body.description,
        system_prompt=body.system_prompt,
        tools=body.tools,
        default_next_mode=body.default_next_mode,
        icon=body.icon,
        color=body.color,
        default_folder=body.default_folder,
        is_builtin=False,
    )
    _save(mode)
    return {"ok": True, "mode": mode.model_dump()}


@modes.router.put("/{mode_id}")
async def update_mode(mode_id: str, body: ModeUpdate):
    mode = _load(mode_id)
    for k, v in body.model_dump(exclude_unset=True).items():
        setattr(mode, k, v)
    _save(mode)
    return {"ok": True, "mode": mode.model_dump()}


@modes.router.post("/{mode_id}/reset")
async def reset_mode(mode_id: str):
    """Reset a built-in mode to its hardcoded defaults."""
    builtin = next((m for m in BUILTIN_MODES if m.id == mode_id), None)
    if not builtin:
        raise HTTPException(status_code=400, detail="Only built-in modes can be reset")
    _save(builtin)
    return {"ok": True, "mode": builtin.model_dump()}


@modes.router.delete("/{mode_id}")
async def delete_mode(mode_id: str):
    mode = _load(mode_id)
    if mode.is_builtin:
        raise HTTPException(status_code=403, detail="Cannot delete built-in modes")
    _store.delete(mode_id)
    return {"ok": True}

File: frontend/src/shared/backend-bridge/apps/modes.ts
import { createAsyncThunk } from '@reduxjs/toolkit';
import { API_BASE } from '@/shared/backend-bridge/base_routes';

const MODES_API: string = `${API_BASE}/modes`;

export interface Mode {
  id: string;
  name: string;
  description: string;
  system_prompt: string | null;
  tools: string[] | null;
  default_next_mode: string | null;
  is_builtin: boolean;
  icon: string;
  color: string;
  default_folder: string | null;
}

// ---------------------------------------------------------------------------
// Mode CRUD
// ---------------------------------------------------------------------------


const list_modes_endpoint: string = `${MODES_API}/list`;
async function list_modes_function(): Promise<{ modes: Mode[]; builtin_defaults: Record<string, Mode> }> {
  const res = await fetch(list_modes_endpoint, {
    method: 'GET',
    headers: { 'Content-Type': 'application/json' },
  });
  const data = await res.json();
  return data as { modes: Mode[]; builtin_defaults: Record<string, Mode> };
}
export const LIST_MODES = createAsyncThunk(
  list_modes_endpoint,
  list_modes_function,
);


const get_mode_endpoint: string = `${MODES_API}/get`;
async function get_mode_function(modeId: string): Promise<Mode> {
  const res = await fetch(`${MODES_API}/${modeId}`, {
    method: 'GET',
    headers: { 'Content-Type': 'application/json' },
  });
  const data = await res.json();
  return data as Mode;
}
export const GET_MODE = createAsyncThunk(
  get_mode_endpoint,
  get_mode_function,
);


const create_mode_endpoint: string = `${MODES_API}/create`;
async function create_mode_function(body: {
  name: string;
  description?: string;
  system_prompt?: string | null;
  tools?: string[] | null;
  default_next_mode?: string | null;
  icon?: string;
  color?: string;
  default_folder?: string | null;
}): Promise<Mode> {
  const res = await fetch(create_mode_endpoint, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  });
  const data = await res.json();
  return data.mode as Mode;
}
export const CREATE_MODE = createAsyncThunk(
  create_mode_endpoint,
  create_mode_function,
);


const update_mode_endpoint: string = `${MODES_API}/update`;
async function update_mode_function(args: {
  modeId: string;
  name?: string;
  description?: string;
  system_prompt?: string | null;
  tools?: string[] | null;
  default_next_mode?: string | null;
  icon?: string;
  color?: string;
  default_folder?: string | null;
}): Promise<Mode> {
  const { modeId, ...updates } = args;
  const res = await fetch(`${MODES_API}/${modeId}`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(updates),
  });
  const data = await res.json();
  return data.mode as Mode;
}
export const UPDATE_MODE = createAsyncThunk(
  update_mode_endpoint,
  update_mode_function,
);


const reset_mode_endpoint: string = `${MODES_API}/reset`;
async function reset_mode_function(modeId: string): Promise<Mode> {
  const res = await fetch(`${MODES_API}/${modeId}/reset`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
  });
  const data = await res.json();
  return data.mode as Mode;
}
export const RESET_MODE = createAsyncThunk(
  reset_mode_endpoint,
  reset_mode_function,
);


const delete_mode_endpoint: string = `${MODES_API}/delete`;
async function delete_mode_function(modeId: string): Promise<string> {
  await fetch(`${MODES_API}/${modeId}`, {
    method: 'DELETE',
    headers: { 'Content-Type': 'application/json' },
  });
  return modeId;
}
export const DELETE_MODE = createAsyncThunk(
  delete_mode_endpoint,
  delete_mode_function,
);


const get_mode_by_id_endpoint: string = `${MODES_API}/get_mode_by_id`;
async function get_mode_by_id_function(modeId: string): Promise<Mode | null> {
  const res = await fetch(get_mode_by_id_endpoint, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ mode_id: modeId }),
  });
  const data = await res.json();
  return data as Mode | null;
}
export const GET_MODE_BY_ID = createAsyncThunk(
  get_mode_by_id_endpoint,
  get_mode_by_id_function,
);
